### PR TITLE
lambda functions example : elastic search

### PR DESCRIPTION
**Pull request by SubashRoul**
Fixes https://github.com/fivetran/app/issues/16688

### CODE REVIEW

File: elastic_search/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+3. Create a zip file
+	`zip -r elastic_search_function.zip *`
+4. Create a Lambda function.
+   
+   For elastic search, execution role must have policy
+	```
+	{
+    	  "Version": "2012-10-17",
+    	  "Statement": [
+        	{
```
Comment (tabalinas) [inline elastic_search/README.md]: replace tabs with spaces in code samples so it's displayed as expected:

image

File: elastic_search/index.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
+/*
+This lambda function fetched search results from elastic search instance
```
Comment (tabalinas) [inline elastic_search/index.js]: fetches?

File: elastic_search/index.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    update(request.state, request.secrets, callback)
+}
+
+async function update(state, secrets, callback) {
+    let response = initializeResponse(state);
+     
+    let insertResponse = fetchSearchResult(state, secrets, 'products', 'product', 'Bag');
+    let deleteResponse = fetchSearchResult(state, secrets, 'products', 'product', 'Puma');
+    
+    insertResponse.then(function(searchResult) {
```
Comment (tabalinas) [inline elastic_search/index.js]: use lamdas instead of functions `() => {}`

File: elastic_search/index.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+exports.handler = (request, context, callback) => {
+    update(request.state, request.secrets, callback)
+}
+
+async function update(state, secrets, callback) {
+    let response = initializeResponse(state);
+     
+    let insertResponse = fetchSearchResult(state, secrets, 'products', 'product', 'Bag');
```
Comment (tabalinas) [inline elastic_search/index.js]: is it case insensitive? we have `bag` and `puma` in comments above

File: elastic_search/index.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    response.state = {
+        "search" : new Date()
+    }
+    
+    callback(null, response);
+}
+
+function initializeResponse(state) {
+    return {
+        "state": {...state},
```
Comment (tabalinas) [inline elastic_search/index.js]: I'm not sure why we enclose js object fields into quotes, it's redundant. 
But if we have this in our docs, we should be at least consistent and add quotes for `hasMore` too.
While quotes are for json format, while here it's js object declaration.

File: elastic_search/index.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            "search": {
+                "primary_key": ['_id', '_index', '_type', 'sync_time']
+            }
+        },
+        hasMore: false
+    };
+}
+
+function fetchSearchResult(state, secrets, indexName, type, query) {
+    let elasticsearch = require('elasticsearch');
```
Comment (tabalinas) [inline elastic_search/index.js]: `require` is usually at the top of the file, so we see all the dependencies.

Comment (tabalinas) [inline elastic_search/index.js]: use `const` for required dependencies

File: elastic_search/index.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    
+    let client = new elasticsearch.Client({
+  	    host: secrets.host,
+  	    log: 'trace'
```
Comment (tabalinas) [inline elastic_search/index.js]: double quotes for consistency

File: elastic_search/index.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    });
+    
+    let a = client.search({
```
Comment (tabalinas) [inline elastic_search/index.js]: why `a`?

File: elastic_search/index.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    
+    return a;
+}
+
+function putSearchResult(response, searchResult, recordOperation) {
+    let index;
+    
+    let hits = searchResult.hits.hits;
+    
+    for(index = 0; index < hits.length; index++) {
```
Comment (tabalinas) [inline elastic_search/index.js]: `hits.forEach`

File: elastic_search/index.js [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        let keys = Object.keys(hits[index]);
+        
+        let keyIndex;
+        let result = {};
+        
+        for(keyIndex = 0; keyIndex < keys.length; keyIndex++) {
```
Comment (tabalinas) [inline elastic_search/index.js]: `let` can be put inside `for`
but I'd rather use `forEach` (unlike in java)

File: elastic_search/index.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+            
+            if(hits[index][key] instanceof Object) {
+                putNestedObject(result, key, hits[index][key]);
+                continue;
+            }
+            
+            result[key] = hits[index][key];
+        }
+        
+        result['sync_time'] = new Date();
```
Comment (tabalinas) [inline elastic_search/index.js]: double quotes (or single everywhere)

File: elastic_search/index.js (hunk 5 above)
Comment (tabalinas) [reply elastic_search/index.js]: btw, no quotes here https://github.com/fivetran/functions/pull/6/files#diff-de246c2522a0583092e324bcb864f070R74 (agree with that)

File: elastic_search/index.js (hunk 4 above)
Comment (SubashRoul) [reply elastic_search/index.js]: Yes, it is case insensitive. But I have changed to `Bag` and `Puma`

**Pull request merged**